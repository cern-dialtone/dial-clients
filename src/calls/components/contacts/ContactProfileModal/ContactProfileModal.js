import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Dimmer,
  Header,
  Icon,
  Loader,
  Modal,
  Segment
} from 'semantic-ui-react';
import { formatUserOrganization } from 'calls/utils/formatters';
import UserPhoneNumberButtonContainer from 'calls/components/UserPhoneNumberButton/UserPhoneNumberButtonContainer';
import ContactAddButton from 'calls/components/contacts/ContactAddButton/ContactAddButton';
import UserProfileExtraInfo from 'calls/components/UserProfileExtraInfo/UserProfileExtraInfo';

function ContactProfileModalContent({ profile }) {
  return (
    <Modal.Content>
      <UserProfileExtraInfo
        mail={profile.mail}
        physicalDeliveryOfficeName={profile.physicalDeliveryOfficeName}
      />
      {profile.phones &&
        profile.phones.map((phone, index) => {
          if (phone.number !== null) {
            return (
              <UserPhoneNumberButtonContainer
                key={`button-${index}`}
                phoneNumber={phone.number}
                icon={phone.phoneType}
                callerName={profile.displayName}
              />
            );
          }
          return null;
        })}
    </Modal.Content>
  );
}

ContactProfileModalContent.propTypes = {
  profile: PropTypes.shape({
    phones: PropTypes.arrayOf.isRequired
  })
};

ContactProfileModalContent.defaultProps = {
  profile: null
};

function ContactProfileModalHeader({ profile }) {
  return (
    <Header>
      <Icon name="user" color="blue" />
      <Header.Content>
        <Header as="h5" floated="right">
          <ContactAddButton contact={profile} />
        </Header>
        {profile ? profile.displayName : ''}
        <Header.Subheader>
          {profile ? formatUserOrganization(profile) : ''}
        </Header.Subheader>
      </Header.Content>
    </Header>
  );
}

ContactProfileModalHeader.propTypes = {
  profile: PropTypes.shape({
    displayName: PropTypes.string
  })
};

ContactProfileModalHeader.defaultProps = {
  profile: {
    displayName: ''
  }
};

export class ContactProfileModal extends Component {
  static propTypes = {
    modalOpen: PropTypes.bool.isRequired,
    unSelectContact: PropTypes.func.isRequired,
    profile: PropTypes.shape({
      number: PropTypes.string
    }),
    fetching: PropTypes.bool.isRequired
  };

  static defaultProps = {
    profile: null
  };

  handleClose = () => {
    const { unSelectContact } = this.props;
    unSelectContact();
  };

  render() {
    const { modalOpen, profile, fetching } = this.props;
    return (
      <Modal
        dimmer="blurring"
        size="tiny"
        open={modalOpen}
        onClose={this.handleClose}
        closeIcon
      >
        {fetching ? (
          <Modal.Content>
            <Segment basic>
              <Dimmer active inverted>
                <Loader inverted size="large" />
              </Dimmer>
            </Segment>
          </Modal.Content>
        ) : (
          <React.Fragment>
            <ContactProfileModalHeader profile={profile} />
            <ContactProfileModalContent profile={profile} />
          </React.Fragment>
        )}
      </Modal>
    );
  }
}

export default ContactProfileModal;
