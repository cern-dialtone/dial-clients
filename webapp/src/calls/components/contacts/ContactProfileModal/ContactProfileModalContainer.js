import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { unSelectContact, getUserProfileById } from "calls/actions/contacts";
import ContactProfileModal from "calls/components/contacts/ContactProfileModal/ContactProfileModal";

function mapStateToProps({ calls }) {
  return {
    modalOpen: calls.contacts.modal.modalOpen,
    selectedContact: calls.contacts.modal.selectedContact
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      unSelectContact,
      getUserProfileById
    },
    dispatch
  );
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(ContactProfileModal);
