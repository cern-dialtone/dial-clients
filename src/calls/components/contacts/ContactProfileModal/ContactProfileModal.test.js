import React from 'react';
import { render, waitForElement } from '@testing-library/react';
import ContactProfileModal from './ContactProfileModal';
import configureMockStore from 'redux-mock-store';
import { createHashHistory } from 'history';
import thunk from 'redux-thunk';
import { routerMiddleware } from 'connected-react-router';
import { Provider } from 'react-redux';

global.remote = {
  require: jest.fn()
};
const history = createHashHistory();
const middlewares = [thunk, routerMiddleware(history)];
const mockStore = configureMockStore(middlewares);
const initialState = {
  contacts: {
    getContacts: {
      contacts: {
        contacts: [
          {
            personId: '1'
          },
          {
            personId: '2'
          },
          {
            personId: '3'
          }
        ]
      }
    },
    addContacts: {
      added: []
    },
    removeContacts: {
      removed: []
    }
  },
  status: {
    doNotDisturb: false
  },
  call: {
    tempRemote: null
  },
  auth: {
    authToken: null
  },
  settings: {
    devices: []
  },
  connection: {
    connected: true
  }
};
const store = mockStore(initialState);

describe('ContactProfileModal tests', () => {
  it('renders the modal closed', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: true,
        incoming: true,
        startTime: 45678
      }
    ];
    const { asFragment, queryByTestId } = render(
      <Provider store={store}>
        <ContactProfileModal
          modalOpen={false}
          profile={{
            personId: '4',
            mail: 'testmail@cern.ch',
            physicalDeliveryOfficeName: 'Name',
            username: 'username',
            phones: [],
            displayName: 'User Name',
            division: 'IT',
            cernGroup: 'CDA',
            cernSecction: 'IC'
          }}
          fetching={false}
          unSelectContact={() => {}}
        />
      </Provider>
    );
    const modal = queryByTestId('ContactProfileModal');
    expect(modal).toBeNull();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the modal open', async () => {
    const recentCalls = [
      {
        name: 'John Doe',
        endTime: 12345,
        phoneNumber: '222333',
        missed: true,
        incoming: true,
        startTime: 45678
      }
    ];
    const { getByTestId } = render(
      <Provider store={store}>
        <ContactProfileModal
          modalOpen={true}
          profile={{
            personId: '4',
            mail: 'testmail@cern.ch',
            physicalDeliveryOfficeName: 'Name',
            username: 'username',
            phones: [],
            displayName: 'User Name',
            division: 'IT',
            cernGroup: 'CDA',
            cernSecction: 'IC'
          }}
          fetching={false}
          unSelectContact={() => {}}
        />
      </Provider>
    );
    await waitForElement(() => getByTestId('ContactProfileModal'));
  });
});
