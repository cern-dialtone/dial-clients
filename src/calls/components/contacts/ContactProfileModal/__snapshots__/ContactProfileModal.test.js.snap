// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactProfileModal tests renders the modal closed 1`] = `<DocumentFragment />`;
